### PR TITLE
Torch-Tensorrt Integration with LightningModule

### LINKED ISSUE #11438: Integrate TorchTensorRt in order to increase speed during inference
## 🚀 Feature

Add a method like to_torchscript in lightgning.py that allow to convert a model in TorchTensorRT in order to increase performance

### Motivation

Increase performance during inference

### Proposal

```py
    @torch.no_grad()
    def to_torch_tensorrt(
            self,
            example_inputs: Optional[Any] = None,
            enabled_precisions: Union ( torch.dtype , torch_tensorrt.dtype
            **kwargs,
    ) -> Union[ScriptModule, Dict[str, ScriptModule]]:
        mode = self.training

        # if no example inputs are provided, try to see if model has example_input_array set
        if example_inputs is None:
            if self.example_input_array is None:
                raise ValueError(
                    "Choosing method=`trace` requires either `example_inputs`"
                    " or `model.example_input_array` to be defined."
                )
            example_inputs = self.example_input_array

        # automatically send example inputs to the right device and use trace
        example_inputs = self._apply_batch_transfer_handler(example_inputs)
        trt_module = torch_tensorrt.compile(self.eval(),
                                           inputs=example_inputs,
                                           enabled_precisions=enabled_precision  # Run with FP16
                                           )
        self.train(mode)

        return trt_module
```

### Additional context

A possible problem could be the dependencies because it depends on  CUDA, cuDNN and TensorRT as you can see https://nvidia.github.io/Torch-TensorRT/v1.0.0/tutorials/installation.html and some of these dependencies I think work only on Linux

cc @borda @carmocca @awaelchli @ninginthecloud @daniellepintz @rohitgr7

### PR DESCRIPTION
**Pull request by GdoongMathew**
## What does this PR do?

Fixes #11438 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by Borda** [comment]
> since there's compatibility problem when testing with the oldest torch 2.1.2 and torch-tensorrt

I think it is fine to test for recent versions, meaning we can skip it, for example, for PT <=2.4 and add a todo note...

File: src/lightning/pytorch/core/module.py [hunk 1]
```diff
@@ -1493,6 +1495,105 @@ def forward(self, x):
...
+            model = SimpleModel()
+            input_sample = torch.randn(1, 64)
+            exported_program = model.to_tensorrt(
+                file_path="export.ep",
+                inputs=input_sample,
+            )
+
+        """
+
+        if not _TORCH_TRT_AVAILABLE:
```
Comment (Borda) [inline src/lightning/pytorch/core/module.py:1621]: Here, you can also raise an exception for some older PT versions if desired

Comment (GdoongMathew) [reply src/lightning/pytorch/core/module.py:1621]: Thanks for the suggestion!! I've added it in https://github.com/Lightning-AI/pytorch-lightning/pull/20808/commits/15f19619247f670815e8e1771c0020d6ef0b0f8b

**Review by Borda — APPROVED** [review]
over all looks good, could you pls check the failing tests?

**Comment by GdoongMathew** [comment]
Hi @Borda, thanks for reviewing this PR!!

 I've managed to fix some of the tests by "not" installing `torch-tensorrt` under certain conditions, but there's still one configuration remaining: `ubuntu-22.04, lightning, 3.12.7, 2.4.1`, and I would likely need some help with this one.

In particular, since `torch-tensorrt` hasn't released its `2.4.x` version for python 3.12, it's impossible for that job to pass. Additionaly, I'm wondering if we could install the library only when a GPU is available since it doesn't really make sense (at least to me) to export in TensorRT format without any GPU in the environment.

**Comment by GdoongMathew** [comment]
I just saw that there's a utility script from https://raw.githubusercontent.com/Lightning-AI/utilities/main/scripts/adjust-torch-versions.py, which would figure out the mapping between `torchvision`, `torchtext` and `torch`, probably I could also add `torch-tensorrt` in the mapping as well ?

But I think extending the mapping only allow the lightning to install `torch-tensorrt` in most of the CICD jobs, but not necessarily solve the dependency missing issue in configuration `python 3.12.7, torch 2.4.1`.

**Comment by GdoongMathew** [comment]
Somehow `torch-tensorrt 2.4.0` for python 3.12 is only available in the pytorch index https://download.pytorch.org/whl/torch-tensorrt, but not in the PyPi.

**Comment by GdoongMathew** [comment]
The failure in the GPU test is due to https://github.com/pytorch/TensorRT/issues/3775

**Comment by GdoongMathew** [comment]
Hi @Borda , all jobs should be fixed now. I also changed a few things at https://github.com/Lightning-AI/pytorch-lightning/pull/20808/commits/c4d2263ebe14ae13a22a6b0443d6e45d95bcbd3b, please let me know if the change makes sense to you. Thanks ~

**Review by Borda — APPROVED** [review]
